### DISCUSSION TITLE
Generating SDK, when Nestjs is blind to any wrapped relations

### DISCUSSION THREAD

**Discussion by JustDenP** [mikro-orm/mikro-orm#6207] [Q&A]
Hello, I want to generate an SDK from my OpenAPI (swagger) docs, but the problem is 
https://mikro-orm.io/docs/usage-with-nestjs#serialization-caveat

For example, here is my user response schema from swagger documentation:
```json
{
  "id": 0,
  "email": "string",
  "password": "string",
  "posts": [],
  "conversations": [],
  "favorites": [],
  "followers": [],
  "followed": []
}
```

And here is an example of my Collection fields inside user entity:
```typescript
@ApiProperty({ type: () => [Post] })
@OneToMany(() => Post, (post) => post.author, {
  orphanRemoval: true,
})
posts = new Collection<Post>(this);

@ManyToMany(() => Conversation, 'users', { owner: true, hidden: true })
conversations = new Collection<Conversation>(this);

@ManyToMany({ hidden: true })
favorites = new Collection<Post>(this);
```

Obviously, with these approach I won't get any corrections from an sdk, and it will require to explicitly add those missing types everytime I work with collections. Is there a way to pass actual types? Thank you

**Comment by B4nan** [comment]
You will need to be more specific, I don't really understand what you are asking for, nor what issues you forsee.

**Comment by JustDenP** [comment]
@B4nan Sorry if my previous question was unclear. I’d like to explain the issue I’m facing in more detail.

### Background:

I’m using NestJS with Mikro-ORM, and I’m trying to generate an SDK from my OpenAPI (Swagger) documentation using tools like openapi-generator. I’m running into an issue with how relation fields in my entities are represented in the generated Swagger schema.

### The Problem:

In Mikro-ORM, relation fields are defined using new Collection<...>(this);. Because of this, Swagger doesn’t properly infer the types of these fields, and they end up being represented as empty arrays without any type information.

For example, in my User entity, I have a posts relation:

```typescript
@OneToMany(() => Post, (post) => post.author, { orphanRemoval: true })
posts = new Collection<Post>(this);
```

But in the generated Swagger schema, the posts field looks like this:

{
  "id": number,
  "posts": [],
}

This means that when I generate the SDK and try to use it on my frontend, I don’t get proper typing for the posts field. So when I do:
```typescript
const user = await sdk.user.getMe();
const firstPostTitle = user.posts[0].title;
```
I get a TypeScript error: `Property 'title' does not exist`

To work around this, I have to manually cast the posts field:
```typescript
const user = await sdk.user.getMe();
const posts: Post[] = user.posts;
const firstPostTitle = posts[0].title;
```
But this isn’t ideal, especially since one of the benefits of generating an SDK is to have proper typing.

### What I’ve Tried:

I tried using the @ApiProperty decorator to explicitly define the type for Swagger:
```typescript
@ApiProperty({ type: () => [Post] })
@OneToMany(() => Post, (post) => post.author, { orphanRemoval: true })
posts = new Collection<Post>(this);
```

But unfortunately, this didn’t change the generated Swagger schema—the posts field still lacks type information.

### What I’m Looking For:

I’m wondering if there’s a way to get Swagger to recognize the types of these relation fields, so that the generated OpenAPI schema includes the proper type definitions. Ideally, I’d like the schema to represent the posts field like this:
```
{
  "id": number,
  "posts": [
    {
      "id": number,
      "author": { /* ... */ },
      "title": string
    }
  ],
}
```

### Question:

Is there a way to configure Mikro-ORM or NestJS so that relation fields defined with new Collection<...>(this); are properly typed in the OpenAPI schema? 

Thanks in advance!

**Reply by B4nan** [reply]
> and I’m trying to generate an SDK from my OpenAPI

And that SDK would be working with your API (not with the ORM), right?

> "posts": [],

I'd guess this can be because of a circular reference. You could rule that out or confirm it by testing a M:N relation on some entity that is not used anywhere else (and doesn't have an inverse side).

> `@ApiProperty({ type: () => [Post] })`

I would say this is the right approach, the nest docs say this is the way to get around cycles too. 

> But unfortunately, this didn’t change the generated Swagger schema—the posts field still lacks type information.

If it doesn't work, it sounds like a swagger (or the nest swagger wrapper) issue, maybe try asking on their end (and be sure to let me know the response, I am also curious). It sounds like they ignore your `type` option you provide explicitly, which to me is a bug on their end. But I never used this approach myself, so take that with a grain of salt, I might be missing something. Havent found anything in their github issues right now.

Either way this does not sound like something to be fixed (or configured) on our end.
